doc/developer/design/20220226_multiwriter_persist.md

```diff
@@ -0,0 +1,206 @@
+
+Initially a persistent collection starts with a merged area at version `0` that
+contains no batches.
+
+### The pending area
+
+Each pending area contains a list of batches produced by a particular writer
+that are contiguous in time.
+
+If a collection is being written to by multiple writers, and therefore there
+are multiple pending areas, then the batches of the two areas must contain
+identical updates for all times `t`. If this invariant is not upheld then the
+contents of the persistent collection are undefined.
+
+Note that there is no requirement that the `lower` and `upper` frontiers of the
+batches in two separate pending areas to match up. Only that the diffs at each
+particular time are identical.
+
+## Collection lifecycle
+
+A persisted collection is managed by the PERSIST CONTROLLER which has the following API:
+
+* `CreateCollection(id)` -> ReadHandle: Creates a new persistent collection
+  identified by `id` and returns a read handle. After a persistent collection
+  has been created anyone can publish new batches into its pending area.
+* `AcquireReadHandle(id) -> ReadHandle`: provides a read handle at its current
+  `since` frontier and referring at the latest merged area version.
+* `DeleteCollection(id)`: Deletes a persistent collection identified by `id`
+
+### Reading
+
+A read handle of a persistent collection is described by a merged area version
+and a compaction frontier. **Readers of a collection must use the batches
+referred to by their current merged area version for any reads at times not
+beyond the merged area's `upper`. For times beyond `upper` the pending areas
+are used instead.**
+
+This behavior is the key to allow the PERSIST CONTROLLER to decide with
+certainty which batches are reachable and perform background garbage collection
+of unreachable batches.
+
+When a reader wants to read at a time `t` that is beyond the merged area's
+`upper` there might be multiple pending areas (due to multiple writers)
+containing a batch that includes that time. Readers are free to choose which
+pending area to read from based on any strategy. For example, a reader could
+pick a pending area to follow and only switch to a different one if its current
+pending area is falling behind too much. The trade-off for switching which
+pending area is followed is the additional work that needs to be done if the
+frontiers of the batches don't exactly line up.
+
+Readers continually try to discover newly minted merged area versions and
+increase their associated version to that, indicating that they will never use
+the batch metadata of any previous version.
+
+Readers are also able to downgrade their compaction frontier indicating that
+they no longer care to distinguish updates that happened not beyond the
+frontier.
+
+Both compaction frontier downgrades and merged area version downgrades are
+durably persisted.
+
+A read handle can be created by cloning an existing read handle. The newly
```

Comment:
I think there may be a race here
- 1 a ReadHandle is coned in-mem and issues an S3 Put to register it in metadata
- 2 the compactor process grabs a list of all registered ReadHandles
- 3 S3 gets the Put for the newly cloned ReadHandle and makes it durable
- 4 all ReadHandles except the newly cloned one advance their since by writing out to the relevant compaction_frontier objects
- 5 the compactor process fetches the compaction_frontier object of each of ReadHandles it found in (2) and makes a compaction decision without knowing about the since of the newly cloned ReadHandle

Reply:
That's a good point, this relies in being able to retrieve the state of all readers in one operation which now that I think about it is not possible in S3. Enforce that readers are registered through the PERSIST CONTROLLER doesn't sound too bad though

Reply:
I've written a bit of prototype code today and the more I think about things, the more I realize that as long as we keep the persist "controller is a singleton process and it must be ensured that only one runs at a given time" requirement then this ends up looking pretty similar to what we already have in src/persist. concretely, imagine if we take src/persist and do the following things

1. remove the ability to have multiple collections in a given persist instance
2. run one persist instance per worker instead of one per process (a decision that long predates platform and is overdue for being revisited). this happens to be completely external to persist
3. optimize a few things internally that are currently threadsafe but don't need to be if we do (2)
4. instead of the current `allow_compaction` cmd, register readers and give them each their own since capability
5. add a new `write_batch` command that takes a set of records and a new upper and the same semantics given here around definiteness and overlapping writes
6. now that we have (5), rip out the current `write` and `seal` cmds. this also makes more sense in a world where we've done (2)
7. introduce something similar to here where writers in processes other than the "persist controller" write batches to s3 in certain locations and the persist controller discovers them and hooks them up to the `write_batch` cmd

After those things, then from the outside, we've arrived back at a persist interface that's interchangeable with what you've described here, modulo that it has lost the "zero coordination" for readers (IIUC not actually a requirement, just the multi writer stuff is, correct me if I'm wrong) and for gc (which we just lost to solve this race condition). However, the correctness (in the sense of bug-free) of that system would be (IMO of course) _far_ easier to reason about than what is here.

On the other hand, as I mentioned, the "single active persist coordinator" requirement has turned out to be a big negative from an operational perspective in a way I didn't originally expect. One thing I'm thinking about a lot right now is if there's some reasonable way to solve that issue by putting a tiny bit of state in etcd/cockroachdb. In that world, my instinct is that what this PR describes is actually closer to what we want than src/persist

Reply:
> this relies in being able to retrieve the state of all readers in one operation which now that I think about it is not possible in S3

Don't want to randomize, but this is possible in s3 if the readers can encode their state in the object name, rather than in the object itself (and we assume less than, say 100 readers)

Reply:
> Don't want to randomize, but this is possible in s3 if the readers can encode their state in the object name, rather than in the object itself (and we assume less than, say 100 readers)

that's true, but I'm not convinced that being able to atomically retrieve the state of all readers is sufficient. I think we need something like compare and set

Reply:
I think registering readers (and maybe writers too) makes sense. If we do that then the race condition is gone because the process that does the registration can be the one providing the overall lower bound for garbage collection frontier since it always knows the full list of readers and can properly synchronize computing a new gc frontier before accepting new reader registrations.

Reply:
A way to lift the need for a single persist coordinator would be to make them compute new versions of the merged area privately and then when they are done attempt to set (in etcd) that for example v2 maps to <uuid>(something that points to their just minted merged area). If this succeeds they were first and every other concurrent compaction process will just generate some private garbage before it figures out that a new compaction process is in charge.

Garbage collection should be safe to run concurrently because all its actions are idempotent (delete an object if it exists) based on some lower bound frontier that can be calculated by looking at the readers in etcd (they can all be read at once)

Reply:
> A way to lift the need for a single persist coordinator would be to make them compute new versions of the merged area privately and then when they are done attempt to set (in etcd) that for example v2 maps to (something that points to their just minted merged area). If this succeeds they were first and every other concurrent compaction process will just generate some private garbage before it figures out that a new compaction process is in charge.
> 
> Garbage collection should be safe to run concurrently because all its actions are idempotent (delete an object if it exists) based on some lower bound frontier that can be calculated by looking at the readers in etcd (they can all be read at once)

Yeah, this is pretty much exactly what I'm thinking and planning to try out in the prototype